Bluetooth: Mesh: redesign key id assignment

alxelax opened the pull request:
The current approach of the key id assignment breaks the key id array consistency after the device rebooting. This PR redesigns the approach of key id management.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
Bluetooth: Mesh: redesign key id assignment

The current approach of the key id assignment breaks
the key id array consistency after the device rebooting.
This PR redesigns the approach of key id management.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -46,6 +46,21 @@ config BT_MESH_USES_MBEDTLS_PSA
 
+if BT_MESH_USES_MBEDTLS_PSA
+
+config BT_MESH_PSA_KEY_ID_USER_MIN_SHIFT
```
Maybe _OFFSET?

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -46,6 +46,21 @@ config BT_MESH_USES_MBEDTLS_PSA
+	help
+	  The PSA specification mandates to set key identifiers for keys
+	  with persistent lifetime. This is responsability of PSA API users
+	  (BLE Mesh is user of PSA API) to provide correct and unique idetifiers.
```
```suggestion
	  (BLE Mesh is user of PSA API) to provide correct and unique identifiers.
```

PavelVPV reviewed (commented):
Looks good. But should CDB be updated since it also stores and restores key ids?

Andrewpini commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -46,6 +46,21 @@ config BT_MESH_USES_MBEDTLS_PSA
+	int "Shift of BLE Mesh key id range regarding PSA_KEY_ID_USER_MIN"
+	default 0
+	help
+	  The PSA specification mandates to set key identifiers for keys
+	  with persistent lifetime. This is responsability of PSA API users
+	  (BLE Mesh is user of PSA API) to provide correct and unique idetifiers.
+	  The BLE Mesh identifier range should be between PSA_KEY_ID_USER_MIN and
+	  PSA_KEY_ID_USER_MAX. BLE Mesh requires 2 ids for each subnetwork plus 2 ids
+	  for each application key and plus 1 id for device key.
```
```suggestion
	  The PSA specification mandates to set key identifiers for keys
	  with persistent lifetime. The users of the PSA API is responsible
	  (BLE Mesh is user of PSA API) to provide correct and unique idetifiers.
	  The BLE Mesh identifier range should be between PSA_KEY_ID_USER_MIN and
	  PSA_KEY_ID_USER_MAX. BLE Mesh requires two ids for each subnetwork, two ids
	  for each application key, and one id for the device key.
```

Andrewpini commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -46,6 +46,21 @@ config BT_MESH_USES_MBEDTLS_PSA
 
+if BT_MESH_USES_MBEDTLS_PSA
+
+config BT_MESH_PSA_KEY_ID_USER_MIN_SHIFT
```
I find it a bit tricky to understand exactly what this parameter is from the description. Do you think it is possible to write it in a more comprehendable way, or is it just that complex?

alxelax replied:
I guess we have a bit different level of view in this. I've already spent significant time to work with PSA specification and changes related to PSA adaptation. Description seems quite clear for me. If you suggest clarifications those help understanding I'll try to integrate them.

alxelax commented:
> Looks good. But should CDB be updated since it also stores and restores key ids?

Seems it should :(. I'll add it too. Then parameter description is not fully relevant since devices with enabled CDB will requiter wider key id range. I'll rewrite description

PavelVPV commented:
Btw, we need to fix includes `shell.c`. I got this warning when compiled `mesh_shell`:
```
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/shell.c: In function 'cmd_net_key_add':
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/shell.c:1165:8: warning: implicit declaration of function 'bt_mesh_key_export'; did you mean 'bt_mesh_key_import'? [-Wimplicit-function-declaration]
 1165 |    if (bt_mesh_key_export(key_val, subnet->keys[0].net_key)) {
      |        ^~~~~~~~~~~~~~~~~~
      |        bt_mesh_key_import
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/shell.c: In function 'cmd_provision':
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/shell.c:2591:3: warning: implicit declaration of function 'bt_mesh_key_destroy'; did you mean 'bt_mesh_key_import'? [-Wimplicit-function-declaration]
 2591 |   bt_mesh_key_destroy(&net_key);
      |   ^~~~~~~~~~~~~~~~~~~
      |   bt_mesh_key_import
```

alxelax committed:
Bluetooth: Mesh: fix shell compilation issues

During key refactorin some shell functions got wrong type of
the incoming variables. The commit fixes it.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV reviewed (approved):
We should probably think later what to do if `CONFIG_BT_MESH_APP_KEY_COUNT` and other configurations reduce after DFU. Since it is not part of composition data, we need to handle in the code a case where we recovered more keys from flash than the firmware allocates.
